client-common/src/tendermint/rpc_client/sync_rpc_client.rs

```diff
@@ -31,32 +47,101 @@ const RESPONSE_TIMEOUT: Duration = Duration::from_secs(10);
 pub struct SyncRpcClient {
     runtime: Arc<Mutex<Runtime>>,
     /// ASYNC RPC CLIENT
-    pub async_rpc_client: AsyncRpcClient,
+    pub async_rpc_client: Arc<Mutex<Option<AsyncRpcClient>>>,
+    url: String,
+}
+
+impl GenesisExt for SyncRpcClient {
+    fn fee_policy(&self) -> LinearFee {
+        static POLICY: OnceCell<LinearFee> = OnceCell::new();
+        let policy = POLICY.get_or_init(|| {
+            self.genesis()
+                .map_err(|e| log::error!("get genesis failed: {:?}", e))
+                .expect("get tendermint genesis")
+                .fee_policy()
+        });
+        policy.clone()
+    }
+}
+
+impl FeeAlgorithm for SyncRpcClient {
+    fn calculate_fee(&self, num_bytes: usize) -> std::result::Result<Fee, CoinError> {
+        self.fee_policy().calculate_fee(num_bytes)
+    }
+
+    fn calculate_for_txaux(&self, txaux: &TxAux) -> std::result::Result<Fee, CoinError> {
+        self.fee_policy().calculate_for_txaux(txaux)
+    }
+}
+
+impl TransactionObfuscation for SyncRpcClient {
```

Comment:
I think this impl can define in client-core, so no need to move cipher from client-core to client-common.

Reply:
`TransactionObfuscation` defined in cipher, but `client-core` needs the `client-common`, so I can't use `client-core` in `client-common`,  if  I do this in client-core, I need to move `SyncRpcClient` to `client-core`

Reply:
yeah, I'll try

Reply:
the `DefaultTransactionObfuscation` and `MockTransactionObfuscation` also located in `client-core/cipher`, but the `client-common` can not use the `client-core`,  so I have to move the cipher to `client-common`

Reply:
I mean you can do this in `client-core`:
```
use client_common::... SyncRpcClient;

impl TransactionObfuscation for SyncRpcClient {
}
```

Because you only need either the trait or the type in current crate to write the impl I think.

Reply:
yeah,   I can impl the trait in the client-core, but the client-common still need the client-core if the client-common wants to use the functions in the trait.

Reply:
ok, but you can move the code, rather than copy, right? I think client-core don't need them anymore.

Reply:
I still think it would be better to not "intrude" into the websocket client implementation itself.
maybe wrap it up with something like this in client-core:
```
struct LazyRpcClient {
    pub rpc_client: Arc<Mutex<Option<SyncRpcClient>>>,
    url: String,
}
```

Reply:
Ok, I'll fix like this.

Reply:
Actually, we can use `LazyRpcClient` as `fee_policy` and `obfuscation`, but we also need the websocket client(which impled the `Client` trait) to initialize the `WalletClient`, so the websocket client should be also `lazy`. Or I have to impl the `Client` trait on the `LazyRpcClient` and use it to create the `WalletClient`, but I don't think it makes sense to create two types of client.